cypress-io/cypress
# feat: Display Cypress Dashboard metrics in the Specs Explorer

- Closes [CLOUD-577](https://cypress-io.atlassian.net/browse/CLOUD-577)

### User facing changelog
* Add new "Latest Runs" and "Average Duration" columns capable of displaying data from the Cypress Dashboard

### Additional details
A large portion of the changes in this PR are to build a GQL mechanism for retrieving & caching Cloud Project data in the app. Whenever possible, previously-cached data is queried and reused by components. As polling proceeds and detects items that may be stale, those individual items can be marked as eligible for re-fetch; once a fetch is initiated they report a status of actively fetching to support display of loading indicators, and finally record a "fetched" or "error" status that can be reflected in the UI.

### How has the user experience changed?
Incorporate metrics from the Cypress Dashboard (cloud) into the app in the form of "latest runs" statistics and "average duration" measurements on a per-spec basis. The intent is to increase the value of logging into the dashboard from the app and expose potential benefits of recording test runs that users may not be aware of. Two new columns are being added to the Specs Explorer as shown in the screenshots below.

**Logged-out State**
For users who are logged out the columns will display disabled placeholders and be otherwise non-interactive.

![Screen Shot 2022-06-26 at 4 42 31 PM](https://user-images.githubusercontent.com/11482842/175834829-8d4081c2-ebc0-477b-ba7f-5f07360f0294.png)

**Activation**
Mousing over the headers will display tooltips describing the action needed to activate these columns. Clicking the login button will start the dashboard login/signup flow and, when complete, will return the user.

![Screen Shot 2022-06-26 at 4 30 38 PM](https://user-images.githubusercontent.com/11482842/175834840-ffff871a-08e4-4462-ab1a-25f81c1eb39e.png)

**Active State**
Once logged-in the Specs Explorer will display the latest statistics (if available) from the cloud. Each column will display a quickly-scannable overview of its measurement to indicate potential problem areas that may need improvement.

![Screen Shot 2022-06-26 at 4 29 10 PM](https://user-images.githubusercontent.com/11482842/175834849-d2724fb6-f40f-46db-b66a-d4dd08d54c66.png)

Each column header displays an explanation of the data being displayed and provides a link to online docs for each.

![Screen Shot 2022-06-26 at 4 29 31 PM](https://user-images.githubusercontent.com/11482842/175834857-705a91d7-22ce-453e-9592-59d3c27deaca.png)

![Screen Shot 2022-06-26 at 4 29 39 PM](https://user-images.githubusercontent.com/11482842/175834863-8d7756fc-a39d-4f7f-8774-e9609565e330.png)

**Near-realtime Updates**
The app will poll for updates to cloud data for visible specs (within the virtualized list or it's overflow rows) - this polling defaults to every 30 seconds but can be remotely accelerated/decelerated by the dashboard as needed to mitigate any load issues. A single request is made to check if the active project has a recording update within the last 30 seconds - if so, queries are launched for each spec to get the latest data for each. This data is reflected in the app by dynamically-updating row components.

In this screenshot, a recording is currently active - some specs at the top have already Passed, one spec is Running, and the remainder are Queued to be run.

![Screen Shot 2022-06-26 at 4 35 11 PM](https://user-images.githubusercontent.com/11482842/175835205-539eeb6b-5017-48a8-b48c-f62a9fce84db.png) 

**Latest Runs Details**
The row items in the "Latest Runs" column gain a mouseover tooltip that exposes additional details on the most recent runs for the given spec.

![Screen Shot 2022-06-26 at 4 29 23 PM](https://user-images.githubusercontent.com/11482842/175834892-c19fc0e5-d60e-4e07-aa0e-c0cb156a7a12.png)

**Not-yet-connected state**
If the user is logged in but the active project has not been connected to the dashboard (has no `projectId` in its config file) the user will be given an actionable header tooltip to connect the project.

![Screen Shot 2022-06-26 at 4 48 37 PM](https://user-images.githubusercontent.com/11482842/175835003-b4eb65c1-e1ad-47df-af4e-0e84f09e7177.png)

**Disconnected State**
If the user is logged in but the active project has an invalid `projectId` the user will be given an actionable header tooltip to re-connect the project.

![Screen Shot 2022-06-26 at 4 31 47 PM](https://user-images.githubusercontent.com/11482842/175835046-692a6104-444b-4bdb-a47f-274d49dff1e1.png)

**Unauthorized State**
If the user is logged in and the project is connected to the dashboard but does not have access to the given `projectId` they will be given an action to request access. This will then update to show "request sent" once actioned.

![Screen Shot 2022-06-26 at 4 32 40 PM](https://user-images.githubusercontent.com/11482842/175835062-582660e8-d6bf-4b1a-8e47-82d0de8155d7.png)

**Offline State**
If a loss of network connectivity is detected we now show a warning banner at the top of the Spec Explorer to notify the user.

![Screen Shot 2022-06-26 at 4 30 17 PM](https://user-images.githubusercontent.com/11482842/175835073-6e494196-c47a-49c3-a0ff-c1fbc54d3fec.png)

### Steps to test
- Confirm all "disconnected" states are correct when logged out
- Log in via the CTAs in the header tooltips
- Confirm all project data is shown correctly after login
- Perform actions such as recording runs, navigation, re-authentication, network throttling, etc.
- Confirm all states are shown and respond to actions correctly

### PR Tasks
- [n/a] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/4554
- [n/a] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [n/a] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

Pre-merge tasks:

- [x] Merge https://github.com/cypress-io/cypress/pull/22347
